src/backend/distributed/planner/merge_planner.c

```diff
@@ -554,27 +625,314 @@ MergeQualAndTargetListFunctionsSupported(Oid resultRelationId, FromExpr *joinTre
+}
+
+
+/*
+ * ConvertRelationRTEIntoSubquery takes a RTE_RELATION and converts it into a RTE_SUBQUERY,
+ * which is basically a SELECT * FROM the relation.
+ */
+static void
+ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
+							   PlannerRestrictionContext *plannerRestrictionContext)
+{
+	Query *sourceResultsQuery = makeNode(Query);
+	RangeTblRef *newRangeTableRef = makeNode(RangeTblRef);
+
+	RelationRestriction *relationRestriction =
+		RelationRestrictionForRelation(sourceRte, plannerRestrictionContext);
+
+	/*
+	 * Hack: RequiredAttrNumbersForRelation uses the planner's query which is
+	 * not capturing the required attributes of the source table from the MERGE
+	 * action list. Here, we temporarily pass the original query to get the
+	 * correct target list of the source rows.
+	 */
+	Query *saveQuery = relationRestriction->plannerInfo->parse;
+	relationRestriction->plannerInfo->parse = mergeQuery;
+	List *requiredAttributes =
+		RequiredAttrNumbersForRelation(sourceRte, plannerRestrictionContext);
+
+	/* Restore the planner's parse tree */
+	relationRestriction->plannerInfo->parse = saveQuery;
+
+	sourceResultsQuery->commandType = CMD_SELECT;
+
+	/* we copy the input rteRelation to preserve the rteIdentity */
+	RangeTblEntry *newRangeTableEntry = copyObject(sourceRte);
+	sourceResultsQuery->rtable = list_make1(newRangeTableEntry);
+
+	/* set the FROM expression to the subquery */
+	newRangeTableRef->rtindex = SINGLE_RTE_INDEX;
+	sourceResultsQuery->jointree = makeFromExpr(list_make1(newRangeTableRef), NULL);
+	sourceResultsQuery->targetList =
+		CreateAllTargetListForRelation(sourceRte->relid, requiredAttributes);
+	List *restrictionList =
+		GetRestrictInfoListForRelation(sourceRte, plannerRestrictionContext);
+	List *copyRestrictionList = copyObject(restrictionList);
+	Expr *andedBoundExpressions = make_ands_explicit(copyRestrictionList);
+	sourceResultsQuery->jointree->quals = (Node *) andedBoundExpressions;
+
+	/*
+	 * Originally the quals were pointing to the RTE and its varno
+	 * was pointing to its index in rtable. However now we converted the RTE
+	 * to a subquery and the quals should be pointing to that subquery, which
+	 * is the only RTE in its rtable, hence we update the varnos so that they
+	 * point to the subquery RTE.
+	 * Originally: rtable: [rte1, current_rte, rte3...]
+	 * Now: rtable: [rte1, subquery[current_rte], rte3...] --subquery[current_rte] refers to its rtable.
+	 */
+	Node *quals = sourceResultsQuery->jointree->quals;
+	UpdateVarNosInNode(quals, SINGLE_RTE_INDEX);
+
+	/* replace the function with the constructed subquery */
+	sourceRte->rtekind = RTE_SUBQUERY;
+	sourceRte->subquery = sourceResultsQuery;
+	sourceRte->inh = false;
+}
+
+
+/*
+ * ConvertSourceRTEIntoSubquery converts MERGE's source RTE into a subquery,
+ * whose result rows are repartitioned during runtime.
+ */
+static void
+ConvertSourceRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
```

Comment:
One another high-level thing that I'm not sure yet and want to get your opinion is here.

Do we really need to update the `mergeQuery` and all the corresponding structure?

Aren't we just trying to calculate the sourceSubquery and then get to `sourcePlan` -- or as you name it in the PR `sourceRowsPlan`? 

Why is that required to change the RTEs on the query etc? Is that really required?

Reply:
I was wondering why I didn't get this error, but my nested CTE is more pipelined in nature

```
WITH cte1 AS (
    SELECT * FROM citus_source ORDER BY 1 LIMIT 90000
),
cte2 AS(
    SELECT * FROM cte1
),
cte3 AS(
    SELECT * FROM cte2
)
MERGE INTO citus_target t
USING cte3 s
ON (s.id=t.id)
WHEN MATCHED AND t.id > 75000 THEN
    UPDATE SET val = s.val + 1
WHEN MATCHED THEN
    DELETE
WHEN NOT MATCHED THEN
    INSERT VALUES (s.id, s.val);

```

Reply:
@onderkalaci  It's on oversight on CTE failing to attach to the source query, fixed it now. Please take a look